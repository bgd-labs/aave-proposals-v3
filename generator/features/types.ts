import {Hex} from 'viem';
import {NumberInputValues, PercentInputValues} from '../prompts';
import {BooleanSelectValues} from '../prompts/boolPrompt';

export interface AssetSelector {
  asset: string;
}

export interface TokenImplementations {
  aToken: Hex;
  vToken: Hex;
  sToken: Hex;
}

export interface CapsUpdatePartial {
  supplyCap: NumberInputValues;
  borrowCap: NumberInputValues;
}

export interface CapsUpdate extends CapsUpdatePartial, AssetSelector {}

export interface BorrowUpdatePartial {
  enabledToBorrow: BooleanSelectValues;
  flashloanable: BooleanSelectValues;
  borrowableInIsolation: BooleanSelectValues;
  withSiloedBorrowing: BooleanSelectValues;
  reserveFactor: PercentInputValues;
}

export interface BorrowUpdate extends BorrowUpdatePartial, AssetSelector {}

export interface CollateralUpdatePartial {
  ltv: PercentInputValues;
  liqThreshold: PercentInputValues;
  liqBonus: PercentInputValues;
  debtCeiling: NumberInputValues;
  liqProtocolFee: PercentInputValues;
}

export interface CollateralUpdate extends CollateralUpdatePartial, AssetSelector {}

export interface PriceFeedUpdatePartial {
  priceFeed: Hex;
}

export interface PriceFeedUpdate extends PriceFeedUpdatePartial, AssetSelector {}

export interface AssetEModeUpdatePartial {
  eModeCategory: string;
  collateral: BooleanSelectValues;
  borrowable: BooleanSelectValues;
}

export interface AssetEModeUpdate extends AssetEModeUpdatePartial, AssetSelector {}

export interface EModeCategoryUpdate {
  // library accessor or new id
  eModeCategory: string | number;
  ltv: NumberInputValues;
  liqThreshold: NumberInputValues;
  liqBonus: NumberInputValues;
  priceSource?: Hex | '';
  label: string;
}

export interface RateStrategyParams {
  optimalUtilizationRate: string;
  baseVariableBorrowRate: string;
  variableRateSlope1: string;
  variableRateSlope2: string;
  stableRateSlope1?: string;
  stableRateSlope2?: string;
  baseStableRateOffset?: string;
  stableRateExcessOffset?: string;
  optimalStableToTotalDebtRatio?: string;
}

export interface RateStrategyUpdate extends AssetSelector {
  params: RateStrategyParams;
}

export interface Listing
  extends CollateralUpdatePartial,
    BorrowUpdatePartial,
    CapsUpdatePartial,
    PriceFeedUpdatePartial {
  asset: Hex;
  assetSymbol: string;
  rateStrategyParams: RateStrategyParams;
  eModeCategory: string;
  decimals: number;
  admin?: Hex | '';
}

export interface ListingWithCustomImpl {
  base: Listing;
  implementations: TokenImplementations;
}

export interface TokenStream {
  asset: Hex;
  receiver: Hex;
  duration: string;
  amount: string;
}

export interface FreezeUpdate extends AssetSelector {
  shouldBeFrozen: boolean;
}

export interface EmissionUpdate {
  asset: Hex;
  symbol: string;
  admin: Hex;
}
