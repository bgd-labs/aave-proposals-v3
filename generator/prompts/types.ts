import {PoolIdentifier} from '../types';

export interface GenericPrompt<T extends boolean = boolean> {
  message: string;
  required?: T;
  transform?: (value: string) => string;
  defaultValue?: string;
}

export interface GenericPoolPrompt<T extends boolean = boolean> extends GenericPrompt<T> {
  pool: PoolIdentifier;
}
