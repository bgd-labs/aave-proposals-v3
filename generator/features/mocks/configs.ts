import {Options} from '../../types';
import {CapsUpdate, Listing} from '../types';

export const MOCK_OPTIONS: Options = {
  pools: ['AaveV3Ethereum'],
  title: 'test',
  shortName: 'Test',
  date: '20231023',
  author: 'test',
  discussion: 'test',
  snapshot: 'test',
};

export const assetListingConfig: Listing[] = [
  {
    assetSymbol: 'PSP',
    decimals: 18,
    priceFeed: '0x72AFAECF99C9d9C8215fF44C77B94B99C28741e8',
    ltv: '40_00',
    liqThreshold: '50_00',
    liqBonus: '5_00',
    debtCeiling: '100_000',
    liqProtocolFee: '20_00',
    enabledToBorrow: 'ENABLED',
    flashloanable: 'ENABLED',
    stableRateModeEnabled: 'DISABLED',
    borrowableInIsolation: 'DISABLED',
    withSiloedBorrowing: 'DISABLED',
    reserveFactor: '20_00',
    supplyCap: '10_000',
    borrowCap: '5_000',
    rateStrategyParams: {
      optimalUtilizationRate: '_bpsToRay(80_00)',
      baseVariableBorrowRate: '_bpsToRay(0_00)',
      variableRateSlope1: '_bpsToRay(10_00)',
      variableRateSlope2: '_bpsToRay(100_00)',
      stableRateSlope1: '_bpsToRay(10_00)',
      stableRateSlope2: '_bpsToRay(100_00)',
      baseStableRateOffset: '_bpsToRay(1_00)',
      stableRateExcessOffset: '_bpsToRay()',
      optimalStableToTotalDebtRatio: '_bpsToRay(10_00)',
    },
    eModeCategory: 'AaveV3EthereumEModes.NONE',
    asset: '0xcAfE001067cDEF266AfB7Eb5A286dCFD277f3dE5',
  },
];
