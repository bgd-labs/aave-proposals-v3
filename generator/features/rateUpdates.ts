import {CodeArtifact, FEATURE, FeatureModule} from '../types';
import {percentInput} from '../prompts';
import {RateStrategyParams, RateStrategyUpdate} from './types';
import {
  assetsSelectPrompt,
  translateAssetToAssetLibUnderlying,
} from '../prompts/assetsSelectPrompt';

export async function fetchRateStrategyParamsV2(
  disableKeepCurrent?: boolean
): Promise<RateStrategyParams> {
  return {
    optimalUtilizationRate: await percentInput({
      message: 'optimalUtilizationRate',
      toRay: true,
      disableKeepCurrent,
    }),
    baseVariableBorrowRate: await percentInput({
      message: 'baseVariableBorrowRate',
      toRay: true,
      disableKeepCurrent,
    }),
    variableRateSlope1: await percentInput({
      message: 'variableRateSlope1',
      toRay: true,
      disableKeepCurrent,
    }),
    variableRateSlope2: await percentInput({
      message: 'variableRateSlope2',
      toRay: true,
      disableKeepCurrent,
    }),
    stableRateSlope1: await percentInput({
      message: 'stableRateSlope1',
      toRay: true,
      disableKeepCurrent,
    }),
    stableRateSlope2: await percentInput({
      message: 'stableRateSlope2',
      toRay: true,
      disableKeepCurrent,
    }),
  };
}

export async function fetchRateStrategyParamsV3(disableKeepCurrent?: boolean) {
  const params = await fetchRateStrategyParamsV2(disableKeepCurrent);
  return {
    ...params,
    baseStableRateOffset: await percentInput({
      message: 'baseStableRateOffset',
      toRay: true,
      disableKeepCurrent,
    }),
    stableRateExcessOffset: await percentInput({
      message: 'stableRateExcessOffset',
      toRay: true,
      disableKeepCurrent,
    }),
    optimalStableToTotalDebtRatio: await percentInput({
      message: 'optimalStableToTotalDebtRatio',
      toRay: true,
      disableKeepCurrent,
    }),
  };
}

export const rateUpdatesV2: FeatureModule<RateStrategyUpdate[]> = {
  value: FEATURE.RATE_UPDATE_V2,
  description: 'RateStrategiesUpdates',
  async cli(opt, pool) {
    console.log(`Fetching information for RatesUpdate on ${pool}`);
    const assets = await assetsSelectPrompt({
      message: 'Select the assets you want to amend',
      pool,
    });
    const response: RateStrategyUpdate[] = [];
    for (const asset of assets) {
      console.log(`Fetching info for ${asset}`);
      response.push({asset, params: await fetchRateStrategyParamsV2()});
    }
    return response;
  },
  build(opt, pool, cfg) {
    const response: CodeArtifact = {
      code: {
        fn: [
          `function rateStrategiesUpdates()
          public
          pure
          override
          returns (IAaveV2ConfigEngine.RateStrategyUpdate[] memory)
        {
          IAaveV2ConfigEngine.RateStrategyUpdate[] memory rateStrategies = new IAaveV2ConfigEngine.RateStrategyUpdate[](${
            cfg.length
          });
          ${cfg
            .map(
              (cfg, ix) => `rateStrategies[${ix}] = IAaveV2ConfigEngine.RateStrategyUpdate({
                asset: ${translateAssetToAssetLibUnderlying(cfg.asset, pool)},
                params: IV2RateStrategyFactory.RateStrategyParams({
                  optimalUtilizationRate: ${cfg.params.optimalUtilizationRate},
                  baseVariableBorrowRate: ${cfg.params.baseVariableBorrowRate},
                  variableRateSlope1: ${cfg.params.variableRateSlope1},
                  variableRateSlope2: ${cfg.params.variableRateSlope2},
                  stableRateSlope1: ${cfg.params.stableRateSlope1},
                  stableRateSlope2: ${cfg.params.stableRateSlope2}
                })
              });`
            )
            .join('\n')}


          return rateStrategies;
        }`,
        ],
      },
    };
    return response;
  },
};

export const rateUpdatesV3: FeatureModule<RateStrategyUpdate[]> = {
  value: FEATURE.RATE_UPDATE_V3,
  description: 'RateStrategiesUpdates',
  async cli(opt, pool) {
    console.log(`Fetching information for RatesUpdate on ${pool}`);
    const assets = await assetsSelectPrompt({
      message: 'Select the assets you want to amend',
      pool,
    });
    const response: RateStrategyUpdate[] = [];
    for (const asset of assets) {
      console.log(`Fetching info for ${asset}`);
      response.push({asset, params: await fetchRateStrategyParamsV3()});
    }
    return response;
  },
  build(opt, pool, cfg) {
    const response: CodeArtifact = {
      code: {
        fn: [
          `function rateStrategiesUpdates()
          public
          pure
          override
          returns (IAaveV3ConfigEngine.RateStrategyUpdate[] memory)
        {
          IAaveV3ConfigEngine.RateStrategyUpdate[] memory rateStrategies = new IAaveV3ConfigEngine.RateStrategyUpdate[](${
            cfg.length
          });
          ${cfg
            .map(
              (cfg, ix) => `rateStrategies[${ix}] = IAaveV3ConfigEngine.RateStrategyUpdate({
                  asset: ${translateAssetToAssetLibUnderlying(cfg.asset, pool)},
                  params: IV3RateStrategyFactory.RateStrategyParams({
                    optimalUsageRatio: ${cfg.params.optimalUtilizationRate},
                    baseVariableBorrowRate: ${cfg.params.baseVariableBorrowRate},
                    variableRateSlope1: ${cfg.params.variableRateSlope1},
                    variableRateSlope2: ${cfg.params.variableRateSlope2},
                    stableRateSlope1: ${cfg.params.stableRateSlope1},
                    stableRateSlope2: ${cfg.params.stableRateSlope2},
                    baseStableRateOffset: ${cfg.params.baseStableRateOffset!},
                    stableRateExcessOffset: ${cfg.params.stableRateExcessOffset!},
                    optimalStableToTotalDebtRatio: ${cfg.params.optimalStableToTotalDebtRatio!}
                  })
                });`
            )
            .join('\n')}


          return rateStrategies;
        }`,
        ],
      },
    };
    return response;
  },
};
