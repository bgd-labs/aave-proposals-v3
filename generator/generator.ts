import fs from 'fs';
import path from 'path';
import {generateContractName, generateFolderName} from './common';
import {proposalTemplate} from './templates/proposal.template';
import {testTemplate} from './templates/test.template';
import {confirm} from '@inquirer/prompts';
import {ConfigFile, Options, PoolConfigs, PoolIdentifier} from './types';
import prettier from 'prettier';
import {generateScript} from './templates/script.template';
import {generateAIP} from './templates/aip.template';

const prettierSolCfg = await prettier.resolveConfig('foo.sol');
const prettierMDCfg = await prettier.resolveConfig('foo.md');
const prettierTsCfg = await prettier.resolveConfig('foo.ts');

type Files = {
  jsonConfig: string;
  script: string;
  aip: string;
  payloads: {payload: string; test: string; contractName: string}[];
};

/**
 * Generates all the file contents for aip/tests/payloads & script
 * @param options
 * @param poolConfigs
 * @returns
 */
export async function generateFiles(options: Options, poolConfigs: PoolConfigs): Promise<Files> {
  const jsonConfig = prettier.format(
    `import {ConfigFile} from '../../generator/types';
    export const config: ConfigFile = ${JSON.stringify({
      rootOptions: options,
      poolOptions: (Object.keys(poolConfigs) as PoolIdentifier[]).reduce((acc, pool) => {
        acc[pool] = {configs: poolConfigs[pool]!.configs, cache: poolConfigs[pool]!.cache};
        return acc;
      }, {}),
    } as ConfigFile)}`,
    {...prettierTsCfg, filepath: 'foo.ts'}
  );

  function createPayloadAndTest(options: Options, pool: PoolIdentifier) {
    const contractName = generateContractName(options, pool);
    const testCode = testTemplate(options, poolConfigs[pool]!, pool);
    return {
      payload: prettier.format(proposalTemplate(options, poolConfigs[pool]!, pool), {
        ...prettierSolCfg,
        filepath: 'foo.sol',
      }),
      test: prettier.format(testCode, {
        ...prettierSolCfg,
        filepath: 'foo.sol',
      }),
      contractName: contractName,
    };
  }

  console.log('generating script');
  const script = prettier.format(generateScript(options), {
    ...prettierSolCfg,
    filepath: 'foo.sol',
  });
  console.log('generating aip');
  const aip = prettier.format(generateAIP(options, poolConfigs), {
    ...prettierMDCfg,
    filepath: 'foo.md',
  });

  return {
    jsonConfig,
    script,
    aip,
    payloads: options.pools.map((pool) => createPayloadAndTest(options, pool)),
  };
}

async function askBeforeWrite(options: Options, path: string, content: string) {
  if (!options.force && fs.existsSync(path)) {
    const force = await confirm({
      message: `A file already exists at ${path} do you want to overwrite`,
      default: false,
    });
    if (!force) return;
  }
  fs.writeFileSync(path, content);
}

/**
 * Writes the files according to defined folder/file format
 * @param options
 * @param param1
 */
export async function writeFiles(options: Options, {jsonConfig, script, aip, payloads}: Files) {
  const baseName = generateFolderName(options);
  const baseFolder = path.join(process.cwd(), 'src', baseName);
  if (fs.existsSync(baseFolder)) {
    if (!options.force && fs.existsSync(baseFolder)) {
      const force = await confirm({
        message: 'A proposal already exists at that location, do you want to continue?',
        default: false,
      });
      if (!force) return;
    }
  } else {
    fs.mkdirSync(baseFolder, {recursive: true});
  }

  // write config
  await askBeforeWrite(options, path.join(baseFolder, 'config.ts'), jsonConfig);
  // write aip
  await askBeforeWrite(options, path.join(baseFolder, `${options.shortName}.md`), aip);
  // write scripts
  await askBeforeWrite(
    options,
    path.join(baseFolder, `${generateContractName(options)}.s.sol`),
    script
  );

  for (const {payload, test, contractName} of payloads) {
    await askBeforeWrite(options, path.join(baseFolder, `${contractName}.sol`), payload);
    await askBeforeWrite(options, path.join(baseFolder, `${contractName}.t.sol`), test);
  }
}
