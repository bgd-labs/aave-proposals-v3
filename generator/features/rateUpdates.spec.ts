// sum.test.js
import {expect, describe, it} from 'vitest';
import {MOCK_OPTIONS, rateUpdateV2} from './mocks/configs';
import {generateFiles} from '../generator';
import {FEATURE, PoolConfigs} from '../types';
import {rateUpdatesV2} from './rateUpdates';

describe('feature: rateUpdatesV2', () => {
  it('should return reasonable code', () => {
    const output = rateUpdatesV2.build({
      options: MOCK_OPTIONS,
      pool: 'AaveV2EthereumAMM',
      cfg: rateUpdateV2,
      cache: {blockNumber: 42},
    });
    expect(output).toMatchSnapshot();
  });

  it('should properly generate files', async () => {
    const poolConfigs: PoolConfigs = {
      ['AaveV2EthereumAMM']: {
        artifacts: [
          rateUpdatesV2.build({
            options: {...MOCK_OPTIONS, pools: ['AaveV2EthereumAMM']},
            pool: 'AaveV2EthereumAMM',
            cfg: rateUpdateV2,
            cache: {blockNumber: 42},
          }),
        ],
        configs: {[FEATURE.RATE_UPDATE_V2]: rateUpdateV2},
        cache: {blockNumber: 42},
      },
    };
    const files = await generateFiles({...MOCK_OPTIONS, pools: ['AaveV2EthereumAMM']}, poolConfigs);
    expect(files).toMatchSnapshot();
  });
});
