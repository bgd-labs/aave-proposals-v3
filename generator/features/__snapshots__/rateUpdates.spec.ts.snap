// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`feature: rateUpdatesV2 > should properly generate files 1`] = `
{
  "aip": "---
title: "test"
author: "test"
discussions: "test"
snapshot: "test"
---

## Simple Summary

## Motivation

## Specification

## References

- Implementation: [AaveV2EthereumAMM](https://github.com/bgd-labs/aave-proposals-v3/blob/main/src/20231023_AaveV2EthereumAMM_Test/AaveV2EthereumAMM_Test_20231023.sol)
- Tests: [AaveV2EthereumAMM](https://github.com/bgd-labs/aave-proposals-v3/blob/main/src/20231023_AaveV2EthereumAMM_Test/AaveV2EthereumAMM_Test_20231023.t.sol)
- [Snapshot](test)
- [Discussion](test)

## Copyright

Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).
",
  "jsonConfig": "import {ConfigFile} from '../../generator/types';
export const config: ConfigFile = {
  rootOptions: {
    pools: ['AaveV2EthereumAMM'],
    title: 'test',
    shortName: 'Test',
    date: '20231023',
    author: 'test',
    discussion: 'test',
    snapshot: 'test',
  },
  poolOptions: {
    AaveV2EthereumAMM: {
      configs: {
        RATE_UPDATE_V2: [
          {
            asset: 'WETH',
            params: {
              optimalUtilizationRate: '',
              baseVariableBorrowRate: '6',
              variableRateSlope1: '',
              variableRateSlope2: '',
              stableRateSlope1: '',
              stableRateSlope2: '',
            },
          },
          {
            asset: 'DAI',
            params: {
              optimalUtilizationRate: '',
              baseVariableBorrowRate: '4',
              variableRateSlope1: '10',
              variableRateSlope2: '',
              stableRateSlope1: '',
              stableRateSlope2: '',
            },
          },
          {
            asset: 'USDC',
            params: {
              optimalUtilizationRate: '',
              baseVariableBorrowRate: '4',
              variableRateSlope1: '10',
              variableRateSlope2: '',
              stableRateSlope1: '',
              stableRateSlope2: '',
            },
          },
          {
            asset: 'USDT',
            params: {
              optimalUtilizationRate: '',
              baseVariableBorrowRate: '6',
              variableRateSlope1: '10',
              variableRateSlope2: '',
              stableRateSlope1: '',
              stableRateSlope2: '',
            },
          },
          {
            asset: 'WBTC',
            params: {
              optimalUtilizationRate: '',
              baseVariableBorrowRate: '5',
              variableRateSlope1: '',
              variableRateSlope2: '',
              stableRateSlope1: '',
              stableRateSlope2: '',
            },
          },
        ],
      },
      cache: {blockNumber: 42},
    },
  },
};
",
  "payloads": [
    {
      "contractName": "AaveV2EthereumAMM_Test_20231023",
      "payload": "// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import {AaveV2EthereumAMMAssets} from 'aave-address-book/AaveV2EthereumAMM.sol';
import {AaveV2PayloadEthereumAMM} from 'aave-helpers/src/v2-config-engine/AaveV2PayloadEthereumAMM.sol';
import {EngineFlags} from 'aave-v3-origin/contracts/extensions/v3-config-engine/EngineFlags.sol';
import {IAaveV2ConfigEngine} from 'aave-helpers/src/v2-config-engine/IAaveV2ConfigEngine.sol';
import {IV2RateStrategyFactory} from 'aave-helpers/src/v2-config-engine/IV2RateStrategyFactory.sol';
/**
 * @title test
 * @author test
 * - Snapshot: test
 * - Discussion: test
 */
contract AaveV2EthereumAMM_Test_20231023 is AaveV2PayloadEthereumAMM {
  function rateStrategiesUpdates()
    public
    pure
    override
    returns (IAaveV2ConfigEngine.RateStrategyUpdate[] memory)
  {
    IAaveV2ConfigEngine.RateStrategyUpdate[]
      memory rateStrategies = new IAaveV2ConfigEngine.RateStrategyUpdate[](5);
    rateStrategies[0] = IAaveV2ConfigEngine.RateStrategyUpdate({
      asset: AaveV2EthereumAMMAssets.WETH_UNDERLYING,
      params: IV2RateStrategyFactory.RateStrategyParams({
        optimalUtilizationRate: EngineFlags.KEEP_CURRENT,
        baseVariableBorrowRate: _bpsToRay(6_00),
        variableRateSlope1: EngineFlags.KEEP_CURRENT,
        variableRateSlope2: EngineFlags.KEEP_CURRENT,
        stableRateSlope1: EngineFlags.KEEP_CURRENT,
        stableRateSlope2: EngineFlags.KEEP_CURRENT
      })
    });
    rateStrategies[1] = IAaveV2ConfigEngine.RateStrategyUpdate({
      asset: AaveV2EthereumAMMAssets.DAI_UNDERLYING,
      params: IV2RateStrategyFactory.RateStrategyParams({
        optimalUtilizationRate: EngineFlags.KEEP_CURRENT,
        baseVariableBorrowRate: _bpsToRay(4_00),
        variableRateSlope1: _bpsToRay(10_00),
        variableRateSlope2: EngineFlags.KEEP_CURRENT,
        stableRateSlope1: EngineFlags.KEEP_CURRENT,
        stableRateSlope2: EngineFlags.KEEP_CURRENT
      })
    });
    rateStrategies[2] = IAaveV2ConfigEngine.RateStrategyUpdate({
      asset: AaveV2EthereumAMMAssets.USDC_UNDERLYING,
      params: IV2RateStrategyFactory.RateStrategyParams({
        optimalUtilizationRate: EngineFlags.KEEP_CURRENT,
        baseVariableBorrowRate: _bpsToRay(4_00),
        variableRateSlope1: _bpsToRay(10_00),
        variableRateSlope2: EngineFlags.KEEP_CURRENT,
        stableRateSlope1: EngineFlags.KEEP_CURRENT,
        stableRateSlope2: EngineFlags.KEEP_CURRENT
      })
    });
    rateStrategies[3] = IAaveV2ConfigEngine.RateStrategyUpdate({
      asset: AaveV2EthereumAMMAssets.USDT_UNDERLYING,
      params: IV2RateStrategyFactory.RateStrategyParams({
        optimalUtilizationRate: EngineFlags.KEEP_CURRENT,
        baseVariableBorrowRate: _bpsToRay(6_00),
        variableRateSlope1: _bpsToRay(10_00),
        variableRateSlope2: EngineFlags.KEEP_CURRENT,
        stableRateSlope1: EngineFlags.KEEP_CURRENT,
        stableRateSlope2: EngineFlags.KEEP_CURRENT
      })
    });
    rateStrategies[4] = IAaveV2ConfigEngine.RateStrategyUpdate({
      asset: AaveV2EthereumAMMAssets.WBTC_UNDERLYING,
      params: IV2RateStrategyFactory.RateStrategyParams({
        optimalUtilizationRate: EngineFlags.KEEP_CURRENT,
        baseVariableBorrowRate: _bpsToRay(5_00),
        variableRateSlope1: EngineFlags.KEEP_CURRENT,
        variableRateSlope2: EngineFlags.KEEP_CURRENT,
        stableRateSlope1: EngineFlags.KEEP_CURRENT,
        stableRateSlope2: EngineFlags.KEEP_CURRENT
      })
    });

    return rateStrategies;
  }
}
",
      "pool": "AaveV2EthereumAMM",
      "test": "// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import {AaveV2EthereumAMM} from 'aave-address-book/AaveV2EthereumAMM.sol';

import 'forge-std/Test.sol';
import {ProtocolV2TestBase, ReserveConfig} from 'aave-helpers/src/ProtocolV2TestBase.sol';
import {AaveV2EthereumAMM_Test_20231023} from './AaveV2EthereumAMM_Test_20231023.sol';

/**
 * @dev Test for AaveV2EthereumAMM_Test_20231023
 * command: FOUNDRY_PROFILE=mainnet forge test --match-path=src/20231023_AaveV2EthereumAMM_Test/AaveV2EthereumAMM_Test_20231023.t.sol -vv
 */
contract AaveV2EthereumAMM_Test_20231023_Test is ProtocolV2TestBase {
  AaveV2EthereumAMM_Test_20231023 internal proposal;

  function setUp() public {
    vm.createSelectFork(vm.rpcUrl('mainnet'), 42);
    proposal = new AaveV2EthereumAMM_Test_20231023();
  }

  /**
   * @dev executes the generic test suite including e2e and config snapshots
   */
  function test_defaultProposalExecution() public {
    defaultTest('AaveV2EthereumAMM_Test_20231023', AaveV2EthereumAMM.POOL, address(proposal));
  }
}
",
    },
  ],
  "scripts": {
    "defaultScript": "// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import {GovV3Helpers, IPayloadsControllerCore, PayloadsControllerUtils} from 'aave-helpers/src/GovV3Helpers.sol';
import {GovernanceV3Ethereum} from 'aave-address-book/GovernanceV3Ethereum.sol';
import {EthereumScript} from 'solidity-utils/contracts/utils/ScriptUtils.sol';
import {AaveV2EthereumAMM_Test_20231023} from './AaveV2EthereumAMM_Test_20231023.sol';

/**
 * @dev Deploy Ethereum
 * deploy-command: make deploy-ledger contract=src/20231023_AaveV2EthereumAMM_Test/Test_20231023.s.sol:DeployEthereum chain=mainnet
 * verify-command: FOUNDRY_PROFILE=mainnet npx catapulta-verify -b broadcast/Test_20231023.s.sol/1/run-latest.json
 */
contract DeployEthereum is EthereumScript {
  function run() external broadcast {
    // deploy payloads
    address payload0 = GovV3Helpers.deployDeterministic(
      type(AaveV2EthereumAMM_Test_20231023).creationCode
    );

    // compose action
    IPayloadsControllerCore.ExecutionAction[]
      memory actions = new IPayloadsControllerCore.ExecutionAction[](1);
    actions[0] = GovV3Helpers.buildAction(payload0);

    // register action at payloadsController
    GovV3Helpers.createPayload(actions);
  }
}

/**
 * @dev Create Proposal
 * command: make deploy-ledger contract=src/20231023_AaveV2EthereumAMM_Test/Test_20231023.s.sol:CreateProposal chain=mainnet
 */
contract CreateProposal is EthereumScript {
  function run() external {
    // create payloads
    PayloadsControllerUtils.Payload[] memory payloads = new PayloadsControllerUtils.Payload[](1);

    // compose actions for validation
    IPayloadsControllerCore.ExecutionAction[]
      memory actionsEthereum = new IPayloadsControllerCore.ExecutionAction[](1);
    actionsEthereum[0] = GovV3Helpers.buildAction(
      type(AaveV2EthereumAMM_Test_20231023).creationCode
    );
    payloads[0] = GovV3Helpers.buildMainnetPayload(vm, actionsEthereum);

    // create proposal
    vm.startBroadcast();
    GovV3Helpers.createProposal(
      vm,
      payloads,
      GovernanceV3Ethereum.VOTING_PORTAL_ETH_POL,
      GovV3Helpers.ipfsHashFile(vm, 'src/20231023_AaveV2EthereumAMM_Test/Test.md')
    );
  }
}
",
  },
}
`;

exports[`feature: rateUpdatesV2 > should return reasonable code 1`] = `
{
  "code": {
    "fn": [
      "function rateStrategiesUpdates()
          public
          pure
          override
          returns (IAaveV2ConfigEngine.RateStrategyUpdate[] memory)
        {
          IAaveV2ConfigEngine.RateStrategyUpdate[] memory rateStrategies = new IAaveV2ConfigEngine.RateStrategyUpdate[](5);
          rateStrategies[0] = IAaveV2ConfigEngine.RateStrategyUpdate({
                asset: AaveV2EthereumAMMAssets.WETH_UNDERLYING,
                params: IV2RateStrategyFactory.RateStrategyParams({
                  optimalUtilizationRate: EngineFlags.KEEP_CURRENT,
                  baseVariableBorrowRate: _bpsToRay(6_00),
                  variableRateSlope1: EngineFlags.KEEP_CURRENT,
                  variableRateSlope2: EngineFlags.KEEP_CURRENT,
                  stableRateSlope1: EngineFlags.KEEP_CURRENT,
                  stableRateSlope2: EngineFlags.KEEP_CURRENT
                })
              });
rateStrategies[1] = IAaveV2ConfigEngine.RateStrategyUpdate({
                asset: AaveV2EthereumAMMAssets.DAI_UNDERLYING,
                params: IV2RateStrategyFactory.RateStrategyParams({
                  optimalUtilizationRate: EngineFlags.KEEP_CURRENT,
                  baseVariableBorrowRate: _bpsToRay(4_00),
                  variableRateSlope1: _bpsToRay(10_00),
                  variableRateSlope2: EngineFlags.KEEP_CURRENT,
                  stableRateSlope1: EngineFlags.KEEP_CURRENT,
                  stableRateSlope2: EngineFlags.KEEP_CURRENT
                })
              });
rateStrategies[2] = IAaveV2ConfigEngine.RateStrategyUpdate({
                asset: AaveV2EthereumAMMAssets.USDC_UNDERLYING,
                params: IV2RateStrategyFactory.RateStrategyParams({
                  optimalUtilizationRate: EngineFlags.KEEP_CURRENT,
                  baseVariableBorrowRate: _bpsToRay(4_00),
                  variableRateSlope1: _bpsToRay(10_00),
                  variableRateSlope2: EngineFlags.KEEP_CURRENT,
                  stableRateSlope1: EngineFlags.KEEP_CURRENT,
                  stableRateSlope2: EngineFlags.KEEP_CURRENT
                })
              });
rateStrategies[3] = IAaveV2ConfigEngine.RateStrategyUpdate({
                asset: AaveV2EthereumAMMAssets.USDT_UNDERLYING,
                params: IV2RateStrategyFactory.RateStrategyParams({
                  optimalUtilizationRate: EngineFlags.KEEP_CURRENT,
                  baseVariableBorrowRate: _bpsToRay(6_00),
                  variableRateSlope1: _bpsToRay(10_00),
                  variableRateSlope2: EngineFlags.KEEP_CURRENT,
                  stableRateSlope1: EngineFlags.KEEP_CURRENT,
                  stableRateSlope2: EngineFlags.KEEP_CURRENT
                })
              });
rateStrategies[4] = IAaveV2ConfigEngine.RateStrategyUpdate({
                asset: AaveV2EthereumAMMAssets.WBTC_UNDERLYING,
                params: IV2RateStrategyFactory.RateStrategyParams({
                  optimalUtilizationRate: EngineFlags.KEEP_CURRENT,
                  baseVariableBorrowRate: _bpsToRay(5_00),
                  variableRateSlope1: EngineFlags.KEEP_CURRENT,
                  variableRateSlope2: EngineFlags.KEEP_CURRENT,
                  stableRateSlope1: EngineFlags.KEEP_CURRENT,
                  stableRateSlope2: EngineFlags.KEEP_CURRENT
                })
              });


          return rateStrategies;
        }",
    ],
  },
}
`;
